File: python-algorithm/leetcode/problem_1041.py
"""1041. Robot Bounded In Circle
https://leetcode.com/problems/robot-bounded-in-circle/
"""


def is_robot_bounded(instructions: str) -> bool:
    dirs = [[0, 1], [0, -1], [-1, 0], [1, 0]]  # N,S,W,E
    p = [0, 0, 0]
    for instruction in instructions:
        d = p[2]
        if instruction == 'G':
            p[0] += dirs[d][0]
            p[1] += dirs[d][1]
        elif d == 0:
            p[2] = 2 if instruction == 'L' else 3
        elif d == 1:
            p[2] = 3 if instruction == 'L' else 2
        elif d == 2:
            p[2] = 1 if instruction == 'L' else 0
        else:
            p[2] = 0 if instruction == 'L' else 1
    if p[2] != 0 or p == [0, 0, 0]:
        return True
    return False

File: java-algorithm/src/main/java/com/leetcode/Problem1041.java
package com.leetcode;

/**
 * https://leetcode.com/problems/robot-bounded-in-circle/
 */
public class Problem1041 {
    public boolean isRobotBounded(String instructions) {
        int[][] dirs = new int[][]{{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int d = 0;
        int x = 0, y = 0;
        for (int i = 0; i < instructions.length(); i++) {
            char inst = instructions.charAt(i);
            if (inst == 'G') { // forward
                x += dirs[d][0];
                y += dirs[d][1];
            } else if (inst == 'L') { // turn left
                d = (d + 3) % 4;
            } else {
                d = (d + 1) % 4;
            }
        }
        return (x == 0 && y == 0) || d != 0;
    }
}
